lib/spack/spack/concretize.py

```diff
@@ -203,46 +202,91 @@ def concretize_version(self, spec):
+            and False represents no change."""
+        architecture = spec.architecture
+        #  represents whatever frontend exists, same if on single arch
+
+        #  methods available to the platform class depends on component
+        #  so created a mapping between component and method name.
+
+        if architecture.concrete:
+            return False
+
+        frontend_os = str(platform.operating_system("fe"))
+        frontend_target = str(platform.target("fe"))
+        default_os = str(platform.operating_system("default_os"))
+        default_target = str(platform.target("default_target"))
+
+        # set any frontend architecture if os or target are frontend
+        if architecture.target == frontend_target:
+            architecture.platform_os = frontend_os
+        if architecture.platform_os == frontend_os:
+            architecture.target = frontend_target
+
+        # set any backend architecture if os or target are backend
+        if not architecture.target:
+            architecture.target = default_target
+        if not architecture.platform_os:
+            architecture.platform_os = default_os
+
+        return True  # spec changed
+
     def concretize_architecture(self, spec):
-        """If the spec is empty provide the defaults of the platform. If the
-        architecture is not a string type, then check if either the platform,
-        target or operating system are concretized. If any of the fields are
-        changed then return True. If everything is concretized (i.e the
-        architecture attribute is a namedtuple of classes) then return False.
-        If the target is a string type, then convert the string into a
-        concretized architecture. If it has no architecture and the root of the
-        DAG has an architecture, then use the root otherwise use the defaults
-        on the platform.
-        """
-        root_arch = spec.root.architecture
+        """If the spec's architecture is already concrete then return False.
+        If the spec is missing the architecture portion of the spec then look
+        through the DAG and find a spec with an architecture. If such a spec
+        exists then assign it to the current spec, otherwise provide the
+        defaults for the system. If the architecture is a Arch object then
+        check if either the platform, target, or operating system are
+        concretized. If any of the fields are changed return True."""
         sys_arch = spack.spec.ArchSpec(spack.architecture.sys_type())
-        spec_changed = False
+        front_arch = spack.spec.ArchSpec(
+                spack.architecture.front_end_sys_type())
+
+        if spec.architecture and spec.architecture.concrete:
+            return False
 
-        if spec.architecture is None:
+        spec_changed = False
+        other_arch_spec = find_spec(spec, lambda x: x.architecture, spec.root)
+        other_arch = other_arch_spec.architecture
+        if not other_arch:
```

Comment:
I think this should be `if not spec.architecture and not other_arch` (otherwise this may overwrite preset properties of `spec.architecture`)

Also, `spec_changed` should be true in this case as well.